### ISSUE TITLE
PeptideIndexer with no cleavage

### ISSUE THREAD

**Issue by zhuchcn** [OpenMS/OpenMS#5830]
Hi, we are trying to do library searching on some custom peptide database and we want to search with no cleavage and and no nterm M clip. We are confident that there are not replicate sequence in our database, but PeptideIndexer assigns quite a lot non-unique matches. See below. Btw, in this example below the FASTA doesn't have decoy seqs, cuz we are just testing it out, and it shouldn't affect our point.

```
Info: using 'no cleavage' as enzyme (obtained from idXML) for digestion.
Peptide identification engine: MSFRAGGER
Enzyme: no cleavage
Info: using 'full' as enzyme specificity (obtained from idXML) for digestion.
Progress of 'Load first DB chunk':
^M-- done [took 0.53 s (CPU), 0.53 s (Wall)] --
Building trie ... done (0s)
Compressing trie to BFS format ...
 done (0s)
Mapping 43624 peptides to ? (unknown number of) proteins.
Searching with up to 3 ambiguous amino acid(s) and 0 mismatch(es)!
Progress of 'Aho-Corasick':
^M  0.00 %               ^M-- done [took 3.52 s (CPU), 0.96 s (Wall)] --
Merge took: 0.00 s (wall), 0.01 s (CPU), 0.00 s (system), 0.01 s (user)
Memory usage (Aho-Corasick): 169 MB (working set delta), 193 MB (peak working set delta)


Aho-Corasick done:
  found 42953 hits for 31739 of 43624 peptides.
Peptide hits passing enzyme filter: 42953
     ... rejected by enzyme filter: 140447
Unmatched peptide: HIAEDADR
Unmatched peptide: SGPSPDPK
Unmatched peptide: GGSDGYGSGR
Unmatched peptide: DNQLNVHGF
Unmatched peptide: EQISENK
Unmatched peptide: TGC(Carbamidomethyl)WRPR
Unmatched peptide: VGLLEPMLR
Unmatched peptide: SGPSPDPK
Unmatched peptide: HSTNYSR
Unmatched peptide: HIAEDADR
Unmatched peptide: TGPQEEK
Unmatched peptide: SIGPQAR
Unmatched peptide: TMILFTMASGR
Unmatched peptide: TTFSSETGR
Unmatched peptide: ...
-----------------------------------
Peptide statistics

  unmatched                : 11885 (27 %)
  target/decoy:
    match to target DB only: 31739 (72 %)
    match to decoy DB only : 0 (0 %)
    match to both          : 0 (0 %)

  mapping to proteins:
    no match (to 0 protein)         : 11885
    unique match (to 1 protein)     : 20858
    non-unique match (to >1 protein): 10881
-----------------------------------
Protein statistics

  total proteins searched: 892741
  matched proteins       : 30097 (30097 new)
  matched target proteins: 30097 (100 %)
  matched decoy proteins : 0 (0 %)
  orphaned proteins      : 0 (all removed)
-----------------------------------
``` 

Looking at the source code, two places look suspicious to me. The first place is when calling `isValidProduct`, `allow_nterm_protein_cleavage` is hard coded as `true`. Why is that and maybe that could be controlled by a command line argument?

https://github.com/OpenMS/OpenMS/blob/9a1ff47a0286830e5044e1c5139774f6e180a5a7/src/openms/source/ANALYSIS/ID/PeptideIndexing.cpp#L128

The second is when calling `tokenize`, the start is assigned with `pos - 2`. So maybe when specificity is full and no cleavage, the `cleavage_positions` could be just `pos`?

https://github.com/OpenMS/OpenMS/blob/e333f629867be464bc2ef31ccc4eb67bfa0a9f05/src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp#L187

**Comment by zhuchcn** [comment]
Just want to follow this up with an example of a spectrum matched to two "proteins".

Original pepXML output by MSFragger:

```xml
<spectrum_query start_scan="17" assumed_charge="2" native_id="controllerType=0 controllerNumber=1 scan=17" spectrum="CPCG0100_1.17.17.2" end_scan="17" index="2" precursor_neutral_mass="956.56287" retention_time_sec="5.70368598">
<search_result>
<search_hit peptide="KPPLMMLK" massdiff="0.00775146484375" calc_neutral_pep_mass="956.5551" peptide_next_aa="H" num_missed_cleavages="0" num_tol_term="2" protein_descr="" num_tot_proteins="1" tot_num_ions="14" hit_rank="1" num_matched_ions="4" protein="CIRC-ENST00000345317.7-E2-E3-E4|33" peptide_prev_aa="-" is_rejected="0">
<search_score name="hyperscore" value="5.165"/>
<search_score name="nextscore" value="5.165"/>
<search_score name="expect" value="2.156503e+00"/>
</search_hit>
</search_result>
</spectrum_query>
```

This spectrum is matched to "CIRC-ENST00000345317.7-E2-E3-E4|33", and the sequence is below:

```
>CIRC-ENST00000345317.7-E2-E3-E4|33
KPPLMMLKHL
```

This does not seem like a correct one, because the trailing 'HL' isn't matched.

This is the record after PeptideIndexer:

```xml
<PeptideHit score="2.156503" sequence="KPPLMMLK" charge="2" aa_before="M [" aa_after="] ]" start="1 0" end="8 7" protein_refs="PH_1292 PH_26670" >
        <UserParam type="string" name="target_decoy" value="target"/>
        <UserParam type="int" name="isotope_error" value="0"/>
        <UserParam type="float" name="hyperscore" value="5.165"/>
        <UserParam type="float" name="nextscore" value="5.165"/>
        <UserParam type="float" name="MS:1001330" value="2.156503"/>
        <UserParam type="string" name="protein_references" value="non-unique"/>
</PeptideHit>
```

The protein hits are updated to 'PH_1292' and 'PH_26670', and their sequences are:

```
>CIRC-ENST00000345317.7-E2-E3-E4|30
MKPPLMMLK
>CIRC-ENST00000345317.7-E2-E3-E4|31
KPPLMMLK
```

PH_1292 is CIRC-ENST00000345317.7-E2-E3-E4|30 and PH_26670 is CIRC-ENST00000345317.7-E2-E3-E4|31

So as you can see one of the match is an exact match from end to end with no cleavage, while the other has the leading M clipped.

**Comment by zhuchcn** [comment]
I know you guys are extremely short on hands, so tweaked it on my forked repo. I created a command line argument to turn off `allow_nterm_protein_cleavage`. You can see the diff below. We tested it on our samples and we now can get ride of all non-unique matches. I'm not 100% sure about what kind of rules are there, but I'm happy to submit this as an PR!

https://github.com/zhuchcn/OpenMS/compare/fd01793...4d4389d

And here is the stats of one of our runs:

```
Aho-Corasick done:                                                                                                                                                          
  found 58791 hits for 58791 of 58791 peptides.                                                                                                                             
Peptide hits passing enzyme filter: 58791                                                                                                                                   
     ... rejected by enzyme filter: 143783                                                                                                                                  
-----------------------------------                                                                                                                                         
Peptide statistics                                                                                                                                                          
                                                                                                                                                                            
  unmatched                : 0 (0 %)                                                                                                                                        
  target/decoy:                                                                                                                                                             
    match to target DB only: 58791 (100 %)                                                                                                                                  
    match to decoy DB only : 0 (0 %)                                                                                                                                        
    match to both          : 0 (0 %)                                                                                                                                        
                                                                                                                                                                            
  mapping to proteins:                                                                                                                                                      
    no match (to 0 protein)         : 0                                                                                                                                     
    unique match (to 1 protein)     : 58791                                                                                                                                 
    non-unique match (to >1 protein): 0 
```

**Comment by cbielow** [comment]
this is highly appreciated! thank you!
Please feel free to open a PR.
(we will most likely need to add/extend a test for this, but let that be my concern -- unless you really really want to collect even more karma points :) :) ).

**Comment by zhuchcn** [comment]
Will open the PR shortly!

> (we will most likely need to add/extend a test for this, but let that be my concern -- unless you really really want to collect even more karma points :) :) ).

Not my top priority/interest but I'm really interested in seeing how tests are managed in this repo! Not so obvious to me so far..

**Comment by zhuchcn** [comment]
PR opened at #5853

**Comment by cbielow** [comment]
fixed in #5853
